dashboard/app/public_json_api.go

```diff
@@ -170,3 +177,141 @@ func GetJSONDescrFor(page interface{}) ([]byte, error) {
 	}
 	return json.MarshalIndent(res, "", "\t")
 }
+
+type coveredBlock struct {
+	FromLine int `json:"from_line"`
+	FromCol  int `json:"from_column"`
+	ToLine   int `json:"to_line"`
+	ToCol    int `json:"to_column"`
+}
+
+type functionCoverage struct {
+	FuncName     string          `json:"func_name"`
+	Instrumented int             `json:"total_blocks"`
+	Blocks       []*coveredBlock `json:"covered_blocks"`
+}
+
+type fileCoverage struct {
+	FilePath  string              `json:"file_path"`
+	Functions []*functionCoverage `json:"functions"`
+}
+
+func writeExtAPICoverageFor(ctx context.Context, w io.Writer, ns, repo string) error {
+	// By default, return the previous month coverage. It guarantees the good numbers.
+	//
+	// The alternative is to return the current month.
+	// The numbers will jump every day, on the 1st date may drop down.
+	tps, err := coveragedb.GenNPeriodsTill(1, civil.DateOf(time.Now()).AddDays(-31), "month")
+	if err != nil {
+		return fmt.Errorf("coveragedb.GenNPeriodsTill: %w", err)
+	}
+	defaultTimePeriod := tps[0]
+	covDBClient := GetCoverageDBClient(ctx)
+	ff, err := coveragedb.MakeFuncFinder(ctx, covDBClient, ns, defaultTimePeriod)
+	if err != nil {
+		return fmt.Errorf("coveragedb.MakeFuncFinder: %w", err)
+	}
+	covCh, errCh := coveragedb.FilesCoverageStream(ctx, covDBClient, ns, defaultTimePeriod)
+
+	gzw := gzip.NewWriter(w)
+	defer gzw.Close()
+
+	jsonHeader := "{\n\t\"files\": [\n"
```

Comment:
I think JSON Lines would work better.
* Our code is simpler -- no need to manually combine a single response from multiple parts.
* The code of the others is simpler -- they also need to be able to process such a large file somehow.

Reply:
I was thinking about it. Two potential problems:
1. How to return repo and commit? We can return in the every json line... Gzip will remove the traffic overhead.
2. How to detect problems on the client side? The headers go first and we'll say 200 Ok. If appengine instance is killed after some json line, it'll be hard for client to understand it is not the end.

With json, if there is no "}\n" line, we have a problem.

Reply:
> How to return repo and commit? 

That's indeed interesting. Maybe we could have some special JSONL record type with this metadata?

In any case, we need to think about the receiver side -- if we generate a single big JSON, they would have to first load the full response in the memory and parse it as a whole.

> How to detect problems on the client side?

There's https://en.wikipedia.org/wiki/Chunked_transfer_encoding and it seems it's natively supported by Go's http server.

Reply:
Morning!
The alternatives(second jsonl blocker):
1. To not stream the gzipped response until we finished building it. It allows to get content-length and return proper headers in case of the error. It makes sense to follow this rule for all the handlers. As a side effect it'll fix our wrong response headers problem.
2. Solve the caching problem first. Gzipping to the bucket will also solve the problem.

Reply:
We agreed on the format in #5591 . One big json will work.
Current size is 66M.
It shouldn't be the problem for modern PCs.